From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 8 Feb 2025 14:21:50 +0100
Subject: [PATCH] Avoid-useless-deque-clear-on


diff --git a/net/minecraft/world/ticks/LevelTicks.java b/net/minecraft/world/ticks/LevelTicks.java
index fbf0d3b808c66e8971c747619f6acf7417af5ef7..4a58125bbdb52aa8b70e47a9df12d12edc3dc157 100644
--- a/net/minecraft/world/ticks/LevelTicks.java
+++ b/net/minecraft/world/ticks/LevelTicks.java
@@ -182,7 +182,7 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
     }
 
     private void cleanupAfterTick() {
-        this.toRunThisTick.clear();
+        //this.toRunThisTick.clear(); // Leaf - This method runs after toRunThisTick is polled so this is always empty
         this.containersToTick.clear();
         this.alreadyRunThisTick.clear();
         this.toRunThisTickSet.clear();
